### PR TITLE
Support for downloading scan resources from an external repository

### PR DESCRIPTION
**Pull request by asafgabai**
- [x] All [tests](https://github.com/jfrog/jfrog-idea-plugin/actions/workflows/test.yml) passed. If this feature is not already covered by the tests, I added new tests.
-----
Added the possibility to download scan resources from a custom repository in Artifactory, instead of JFrog's repository.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -54,7 +54,11 @@ def idePluginsCommonVersion = '2.3.0'
 dependencies {
     implementation group: 'com.fasterxml.jackson.dataformat', name: 'jackson-dataformat-yaml', version: '2.15.2'
     implementation group: 'org.jfrog.buildinfo', name: 'build-info-extractor', version: buildInfoVersion
-    implementation group: 'com.jfrog.ide', name: 'ide-plugins-common', version: idePluginsCommonVersion
+    implementation('com.jfrog.ide:ide-plugins-common') {
```
Comment (asafgabai) [inline build.gradle:57]: Reminder for myself to update this before it's merged.

File: src/main/java/com/jfrog/ide/idea/configuration/ServerConfigImpl.java [hunk 2]
```diff
@@ -451,6 +466,18 @@ public void readConnectionDetailsFromEnv() {
...
 
+    /**
+     * Read missing configuration from environment variables.
+     */
+    public void readMissingConfFromEnv() {
+        if (isBlank(getExternalResourcesRepo())) {
+            String externalResourcesRepoEnv = EnvironmentUtil.getValue(EXTERNAL_RESOURCES_REPO_ENV);
+            if (isNotBlank(externalResourcesRepoEnv)) {
+                setExternalResourcesRepo(externalResourcesRepoEnv);
+            }
```
Comment (Or-Geva) [inline src/main/java/com/jfrog/ide/idea/configuration/ServerConfigImpl.java:477]:
```suggestion
          setExternalResourcesRepo(EnvironmentUtil.getValue(EXTERNAL_RESOURCES_REPO_ENV));
```

File: src/main/java/com/jfrog/ide/idea/scan/ScanManager.java [hunk 3]
```diff
@@ -76,7 +76,7 @@ public void startScan() {
             return;
         }
 
-        if (!GlobalSettings.getInstance().reloadXrayCredentials()) {
+        if (!GlobalSettings.getInstance().reloadMissingConfiguration()) {
```
Comment (Or-Geva) [inline src/main/java/com/jfrog/ide/idea/scan/ScanManager.java:79]: Isn't the `reloadXrayCredentials` function call needed? .
If not, let's change the error message on line 80

File: src/main/java/com/jfrog/ide/idea/ui/configuration/JFrogGlobalConfiguration.java [hunk 4]
```diff
@@ -274,6 +284,12 @@ private void loadConfig() {
             watches.setText(serverConfig.getWatches());
             connectionRetries.setValue(serverConfig.getConnectionRetries());
             connectionTimeout.setValue(serverConfig.getConnectionTimeout());
+            if (!StringUtils.isEmpty(serverConfig.getExternalResourcesRepo())) {
+                useExternalRepositoryCheckBox.setSelected(true);
+                repositoryNameJBTextField.setText(serverConfig.getExternalResourcesRepo());
+            } else {
+                useExternalRepositoryCheckBox.setSelected(false);
+            }
```
Comment (Or-Geva) [inline src/main/java/com/jfrog/ide/idea/ui/configuration/JFrogGlobalConfiguration.java:292]: We can remove the select box and have only the text box. It complicates the code. Here, we need to check if there is a text and then set the checkbox accordingly.

**Review by Or-Geva — CHANGES_REQUESTED** [review]

File: src/main/java/com/jfrog/ide/idea/configuration/ServerConfigImpl.java (hunk 2 above)
Comment (asafgabai) [reply src/main/java/com/jfrog/ide/idea/configuration/ServerConfigImpl.java:477]: I did it to handle a case when the value of externalResourcesRepoEnv is an empty string.

File: src/main/java/com/jfrog/ide/idea/ui/configuration/JFrogGlobalConfiguration.java (hunk 4 above)
Comment (asafgabai) [reply src/main/java/com/jfrog/ide/idea/ui/configuration/JFrogGlobalConfiguration.java]: It was changed to radio buttons. For a better UX, we decided to add these radio buttons in addition to the textbox, to make it clearer to the user what happens if the textbox remains empty.

File: src/main/java/com/jfrog/ide/idea/ui/configuration/JFrogGlobalConfiguration.form [hunk 5]
```diff
@@ -578,47 +551,122 @@
...
+          <component id="a9c14" class="com.intellij.ui.components.JBLabel" binding="pluginResourcesDescJBLabel">
+            <constraints>
+              <grid row="11" column="0" row-span="1" col-span="4" vsize-policy="3" hsize-policy="3" anchor="8" fill="0" indent="2" use-parent-layout="false"/>
+            </constraints>
+            <properties>
+              <focusCycleRoot value="true"/>
+              <focusable value="false"/>
+              <font size="11"/>
+              <foreground color="-6513248"/>
+              <text value="&lt;html&gt;&#10;The plugin requires some of its resources to be downloaded.&#10;&lt;br&gt;&#10;Here you can choose where these resources will be downloaded from.&#10;&lt;br&gt;&#10;&lt;a href=&quot;https://www.jetbrains.com/toolbox-app/?utm_source=productutm_medium=linkutm_campaign=toolbox_app_in_IDE_updatewindowutm_content=we_recommend&quot;&gt;Read more&lt;icon src=&quot;AllIcons.Ide.External_link_arrow&quot;&gt;&lt;/a&gt;&#10;&lt;/html&gt;&#10;"/>
```
Comment (asafgabai) [inline src/main/java/com/jfrog/ide/idea/ui/configuration/JFrogGlobalConfiguration.form:637]: Reminder for myself to update the link when the documentation is ready.

File: src/test/java/com/jfrog/ide/idea/integration/ExternalResourcesRepoIntegrationTests.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+            // An exception should've been thrown. If not, the test failed.
+            fail();
+        } catch (FileNotFoundException e) {
+        } catch (Throwable e) {
+            // If an exception other than FileNotFoundException was thrown, then the test failed.
+            fail();
+        } finally {
+            // Restore the original ServerConfig in GlobalSettings
+            GlobalSettings.getInstance().setServerConfig(originalServerConfig);
+        }
```
Comment (Or-Geva) [inline src/test/java/com/jfrog/ide/idea/integration/ExternalResourcesRepoIntegrationTests.java:84]: Can we move the two fail() and use return instead?

**Review by Or-Geva — CHANGES_REQUESTED** [review]
We can add to the GHA yaml the JFROG_IDE_TEST_EXTERNAL_RESOURCES_REPO env var as plain text.

Make sure that we also update https://github.com/jfrog/documentation with the new readme changes.

File: src/test/java/com/jfrog/ide/idea/integration/ExternalResourcesRepoIntegrationTests.java (hunk 6 above)
Comment (yahavi) [inline src/test/java/com/jfrog/ide/idea/integration/ExternalResourcesRepoIntegrationTests.java:84]:
```suggestion
        assertThrows(FileNotFoundException.class, () -> scanner.execute(input, this::dummyCheckCanceled));
        // Restore the original ServerConfig in GlobalSettings
        GlobalSettings.getInstance().setServerConfig(originalServerConfig);
```

File: src/test/java/com/jfrog/ide/idea/integration/ExternalResourcesRepoIntegrationTests.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        ScanConfig.Builder input = new ScanConfig.Builder()
+                .roots(List.of(testProjectRoot));
+        List<JFrogSecurityWarning> results = scanner.execute(input, this::dummyCheckCanceled);
+        assertEquals(8, results.size());
+
+        // Restore the original ServerConfig in GlobalSettings
+        GlobalSettings.getInstance().setServerConfig(originalServerConfig);
+    }
+
+    public void testDownloadScannersFromExternalRepoNotExist() throws IOException, InterruptedException {
```
Comment (yahavi) [inline src/test/java/com/jfrog/ide/idea/integration/ExternalResourcesRepoIntegrationTests.java:60]:
```suggestion
    public void testDownloadScannersFromExternalRepoNotExist() throws IOException {
```

File: src/test/java/com/jfrog/ide/idea/integration/ExternalResourcesRepoIntegrationTests.java [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+import com.jfrog.ide.idea.log.Logger;
+import com.jfrog.ide.idea.scan.ScanBinaryExecutor;
+import com.jfrog.ide.idea.scan.SecretsScannerExecutor;
+import com.jfrog.ide.idea.scan.data.ScanConfig;
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.lang.StringUtils;
+import org.junit.Test;
+import org.mockito.MockedStatic;
+import org.mockito.Mockito;
+import org.mockito.Spy;
```
Comment (yahavi) [inline src/test/java/com/jfrog/ide/idea/integration/ExternalResourcesRepoIntegrationTests.java:16]: Please reorganize the imports

File: src/test/java/com/jfrog/ide/idea/integration/ExternalResourcesRepoIntegrationTests.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    protected void setUp() throws Exception {
+        super.setUp();
+        scanner = new SecretsScannerExecutor(Logger.getInstance());
+    }
+
+    public void testDownloadScannersFromExternalRepo() throws IOException, InterruptedException {
+        String externalResourcesRepo = System.getenv(ENV_EXTERNAL_RESOURCES_REPO);
+        if (StringUtils.isEmpty(externalResourcesRepo)) {
+            fail("The " + ENV_EXTERNAL_RESOURCES_REPO + " environment variable must be set to run this test");
+        }
```
Comment (yahavi) [inline src/test/java/com/jfrog/ide/idea/integration/ExternalResourcesRepoIntegrationTests.java:41]:
```suggestion
        assertFalse("The " + ENV_EXTERNAL_RESOURCES_REPO + " environment variable must be set to run this test", StringUtils.isEmpty(externalResourcesRepo));
```

File: src/test/java/com/jfrog/ide/idea/integration/ExternalResourcesRepoIntegrationTests.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+import com.intellij.util.EnvironmentUtil;
+import com.jfrog.ide.idea.configuration.GlobalSettings;
+import com.jfrog.ide.idea.configuration.ServerConfigImpl;
+import com.jfrog.ide.idea.inspections.JFrogSecurityWarning;
+import com.jfrog.ide.idea.log.Logger;
+import com.jfrog.ide.idea.scan.ScanBinaryExecutor;
+import com.jfrog.ide.idea.scan.SecretsScannerExecutor;
+import com.jfrog.ide.idea.scan.data.ScanConfig;
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.lang.StringUtils;
```
Comment (yahavi) [inline src/test/java/com/jfrog/ide/idea/integration/ExternalResourcesRepoIntegrationTests.java:12]:
```suggestion
import org.apache.commons.lang3.StringUtils;
```

File: src/main/java/com/jfrog/ide/idea/ui/configuration/JFrogGlobalConfiguration.java [hunk 11]
```diff
@@ -22,6 +19,7 @@
 import com.jfrog.xray.client.impl.XrayClientBuilder;
 import com.jfrog.xray.client.impl.util.JFrogInactiveEnvironmentException;
 import com.jfrog.xray.client.services.system.Version;
+import org.apache.commons.lang.StringUtils;
```
Comment (yahavi) [inline src/main/java/com/jfrog/ide/idea/ui/configuration/JFrogGlobalConfiguration.java:22]:
```suggestion
import org.apache.commons.lang3.StringUtils;
```

**Review by yahavi — CHANGES_REQUESTED** [review]

File: src/main/java/com/jfrog/ide/idea/scan/ScanBinaryExecutor.java [hunk 12]
```diff
@@ -192,38 +191,44 @@ protected List<JFrogSecurityWarning> execute(ScanConfig.Builder inputFileBuilder
...
+            try (ArtifactoryManager artifactoryManager = artifactoryManagerBuilder.build()) {
+                if (Files.exists(binaryTargetPath)) {
+                    // Check for new version of the binary
+                    try (FileInputStream archiveBinaryFile = new FileInputStream(archiveTargetPath.toFile())) {
+                        String latestBinaryChecksum = getFileChecksumFromServer(artifactoryManager, externalResourcesRepo);
+                        String currentBinaryCheckSum = DigestUtils.sha256Hex(archiveBinaryFile);
+                        if (latestBinaryChecksum.equals(currentBinaryCheckSum)) {
+                            nextUpdateCheck = currentTime.plusDays(UPDATE_INTERVAL);
+                            return;
+                        } else {
```
Comment (yahavi) [inline src/main/java/com/jfrog/ide/idea/scan/ScanBinaryExecutor.java:210]: Else after return can be removed

**Pull request merged**